Enable RFC 7919 FFDHE groups for TLS 1.2 server

jvdsn opened the pull request:
For TLS versions prior to TLS 1.3, the SSL server currently only uses
the supported_groups extension for ECDHE cipher suites. This commit
extends this functionality by also considering the supported_groups for
DHE cipher suites.

Before this commit, the logic for generating a temporary DH key for DHE
cipher suites is the following:
1) If dh_tmp_auto is set (see SSL_set_dh_auto), the SSL server
   automatically generates a set of DH parameters (P and G) appropriate
   for the security level of the cipher suite.
2) Otherwise, if the user provided a pre-generated set of DH parameters
   (SSL_set0_tmp_dh_pkey), those parameters are used.
3) Finally, if neither 1) or 2) are applicable, a callback function can
   be set using SSL_set_tmp_dh_callback, which will be invoked to
   generate the temporary DH parameters. From OpenSSL 3.0, this
   functionality is deprecated.
4) Using the parameters from step 1-3, an ephemeral DH key is
   generated. The parameters and the public key are sent to the client.

This commit adds an additional step, prior to step 1:
0) If tls1_shared_group returns any shared known group between the
   server and the client, the DH parameters associated with this group
   are selected.

This change also has the added benefit that SSL_set1_groups and friends
now behave as expected for FFDHE groups on the server side (and by
extension, `s_server -groups` too). This provides the user with more
control over which DH parameters are used by the server for key
exchange, which is useful for testing purposes.

Note however, that the new behavior is technically not compliant with
RFC 7919 (Section 4):
  If a compatible TLS server receives a Supported Groups extension from
  a client that includes any FFDHE group (i.e., any codepoint between
  256 and 511, inclusive, even if unknown to the server), and if none
  of the client-proposed FFDHE groups are known and acceptable to the
  server, then the server MUST NOT select an FFDHE cipher suite.

Nevertheless, for backwards-compatibility purposes, a slight deviation
from the RFC is acceptable here.

Finally, it doesn't seem like in TLS < 1.3, the selected FFDHE group is
explicitly sent from the server to the client. In other words, the
server knows that it selected an RFC 7919 FFDHE group, but the client
only receives the DH parameters. Of course, the client could go through
the tedious effort of comparing these parameters with the RFC 7919
FFDHE group parameters to determine which group was selected.

Partially fixes #10971.

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

jvdsn commented:
One open question: do we want to add FFDHE group support to the TLS 1.2 client too? The main reason I didn't include it here is because of the following text in the RFC:
>    o  If the signature over ServerDHParams is valid, compare the
>       selected dh_p and dh_g with the FFDHE groups offered by the
>       client.  If none of the offered groups match, the server is not
>       compatible with this document.  The client MAY decide to continue
>       the connection if the selected group is acceptable under local
>       policy, or it MAY decide to terminate the connection with a fatal
>       insufficient_security(71) alert.

Right now, the OpenSSL client implementation doesn't do any comparison and simply accepts what the server sends.

If we want to be fully compliant with the RFC, that comparison would need to be implemented and there would have to be some logic to handle the case where the server doesn't send an FFDHE group when requested by the client. Currently, every OpenSSL-based TLS 1.2 server out in the wild doesn't send FFDHE groups, even when requested by the client (after all, that's the exact issue this PR is addressing)...

On the other hand, if we don't care about that statement in the RFC, then it would be trivial to add FFDHE group support to the client. Currently, `ssl/statem/extensions_clnt.c` contains the following code:
```
    /*
     * We only support EC groups in TLSv1.2 or below, and in DTLS. Therefore
     * if we don't have EC support then we don't send this extension.
     */
    if (!use_ecc(s, min_version, max_version)
            && (SSL_CONNECTION_IS_DTLS(s) || max_version < TLS1_3_VERSION))
        return EXT_RETURN_NOT_SENT;
```

The `max_version < TLS1_3_VERSION` could simply be removed to ensure FFDHE groups are always sent (except for DTLS).

jvdsn commented:
As a side note, the server-side behavior can be tested in three ways:
1) `openssl s_server -www -cert ... -key ... -tls1_2 --cipher TLS_DHE_RSA_WITH_AES_256_GCM_SHA384 --groups ffdhe8192` then `curl -vvv https://localhost:4433/` -> verify that the server selects ffdhe8192 because that is the only group supported by the server.
2) `openssl s_server -www -cert ... -key ...` then `wget -O - --cipher "NONE:+CIPHER-ALL:+DHE-RSA:+MAC-ALL:+COMP-ALL:+VERS-TLS1.2:+SIGN-ALL:+GROUP-FFDHE8192" https://localhost:4433` -> verify that the server selects ffdhe8192 because that is the only group supported by the client.
3) `openssl s_server -www -cert ... -key ... -tls1_2 --cipher TLS_DHE_RSA_WITH_AES_256_GCM_SHA384 --groups ffdhe8192` then `wget -O - --cipher "NONE:+CIPHER-ALL:+DHE-RSA:+MAC-ALL:+COMP-ALL:+VERS-TLS1.2:+SIGN-ALL:+GROUP-FFDHE4096" https://localhost:4433` -> verify that the server selects auto-generated 2048-bit DH parameters because there are no shared supported groups.

davidben commented:
Note that these ciphers are soon to be marked as deprecated by the IETF. There are a whole host of reasons why one should not use the TLS 1.2 incarnation of FFDHE.
https://www.ietf.org/archive/id/draft-ietf-tls-deprecate-obsolete-kex-03.html
https://mailarchive.ietf.org/arch/msg/tls/4n0EFQg0FtMKDptGnTkzMbmPxpk/
https://mailarchive.ietf.org/arch/msg/tls/1xZt0rxq1QSzJVdjWRY-bYjNbiI/

jvdsn commented:
> Note that these ciphers are soon to be marked as deprecated by the IETF. There are a whole host of reasons why one should not use the TLS 1.2 incarnation of FFDHE. https://www.ietf.org/archive/id/draft-ietf-tls-deprecate-obsolete-kex-03.html https://mailarchive.ietf.org/arch/msg/tls/4n0EFQg0FtMKDptGnTkzMbmPxpk/ https://mailarchive.ietf.org/arch/msg/tls/1xZt0rxq1QSzJVdjWRY-bYjNbiI/

Thanks for the link. If this draft gets published by IETF, perhaps OpenSSL should consider marking all those cipher suites (TLS_RSA_WITH_*, TLS_DH_*_WITH_*, TLS_DHE_*_WITH_*, TLS_ECDH_*_WITH_*) as deprecated. However, until those cipher suites are deprecated by OpenSSL, is there any security concern with using ffdhe* over the automatically generated DH parameters?

kroeckx commented:
Are there any other implementations that support this, but don't support TLS 1.3?

jvdsn commented:
> Are there any other implementations that support this, but don't support TLS 1.3?

Not sure. In terms of support, it looks like there's a couple of GnuTLS releases that included RFC 7919 before finalizing TLS 1.3 (3.6.0-3.6.5). OpenSSL and NSS both added it as part of/after TLS 1.3 (version 1.1.1 and 3.25 respectively). However, in terms of TLS 1.3 as the default, NSS only started doing that in 3.48. Those are all pretty old versions by now.

jvdsn commented:
@t8m I have just pushed a new version that resolves the merge conflicts. Would it be possible to take another look at this? I know I haven't been timely with following up on this PR but I'd like to see a resolution so I know if I should spend more effort on it. If the project decides this is not necessary I'll just close the PR.

The reason why I am pushing for this change is because of the standards set by the [National Information Assurance Partnership (NIAP)](https://www.niap-ccevs.org/), which manages Common Criteria certifications in the US. NIAP publishes standards for TLS implementations ([PKG_TLS_V2.0](https://www.niap-ccevs.org/protectionprofiles/465), both clients and servers), and one of their requirements is the following (here, TSF refers to the TLS server):

> The TSF shall perform key establishment for TLS using [selection:
> RSA with size [selection: 2048, 3072, 4096 ] bits and no other sizes
> Diffie-Hellman parameters with size [selection: 2048, 3072, 4096, 6144,
> 8192 ] bits and no other sizes
> Diffie-Hellman groups [selection: ffdhe2048, ffdhe3072, ffdhe4096,
> ffdhe6144, ffdhe8192 ] and no other groups, consistent with the client's
> supported groups extension and [selection: key share, no other ] extension
> ECDHE parameters using elliptic curves [selection: secp256r1, secp384r1,
> secp521r1 ] and no other curves, consistent with the client's supported
> groups extension and [selection: key share, no other ] extension and using
> non-compressed formatting for points
> ].

In other words, the TLS server could claim compliance to RSA with the specified sizes, non-standard DH parameters the specified sizes, the "safe prime" groups, and the defined elliptic curves. A TLS 1.2 server that relies on OpenSSL currently cannot claim compliance to the "safe prime" groups because of the following test requirement:

> Test 25.3.2: [conditional] If the TSF supports DHE ciphersuites and supports DHE
> groups in the supported groups extension, then for each supported DHE group,
> the evaluator shall send the TSF a compliant TLS 1.2 client hello indicating a
> single ciphersuite that is compatible with the group and indicating the group in
> the supported groups extension. The evaluator shall observe that the TSF
> negotiates TLS 1.2 using the indicated ciphersuite and that the server key
> exchange message indicates the specific group. The evaluator shall send the TOE
> a client key exchange with a valid point in the group and observe that the TSF
> successfully completes the session.

The test requires the evaluator to verify that the TLS server uses the specific group (i.e., ffdheXXXX). OpenSSL does not support this behavior hence the test cannot pass.

In [version 2.1 of this standard](https://commoncriteria.github.io/tls/master/tls-release.html) (currently under development), the non-standard DH parameters will be disallowed completely, and the only DH option will be:

> Diffie-Hellman groups [selection: ffdhe3072, ffdhe4096, ffdhe6144, ffdhe8192] and no other groups, consistent with the client's supported groups extension and [selection: key share, no other] extension 

This would mean there would be no way to claim compliance to any TLS 1.2 DHE ciphersuites using the OpenSSL server, unless OpenSSL supports RFC 7919.

beldmit commented:
Coding style issue is relevant

jvdsn commented:
> Coding style issue is relevant

Yeah I was looking for input on the technical changes first before fixing that. I believe the style (indentation) issue was already present so fixing it would destroy the git diff.

> no... that shouldn't be happening... the OpenSSL server should select one of the well known parameters sets based on the RSA key size it has available.

I did not change that part of the process, are you saying it should be changed?

> Regarding enforcement: I think we can simply have ability to request but not require RFC 7919 groups, and a way to require them (abort if the connection would end up with a DHE cipher but no shared ffdhe groups are supported).

That would be on the client side. OpenSSL currently does have the ability to request these groups and indeed does not require them. If this PR would be accepted I would also look at implementing an optional enforcement (since this is also a test performed by the aforementioned government standard).

jvdsn commented:
In general I am wondering if the RFC 7919 primes should be selected after what I call "step 1" and "step 2" above instead of before. That way the user could still choose not to use them.

tomato42 commented:
> I did not change that part of the process, are you saying it should be changed?

no, I'm saying that this is incorrect description from the code behaviour we have tests for... what I described is the current behaviour of the code as far as I understand it

> That would be on the client side. OpenSSL currently does have the ability to request these groups and indeed does not require them. If this PR would be accepted I would also look at implementing an optional enforcement (since this is also a test performed by the aforementioned government standard).

no... I'd say that it can apply to server behaviour too: refusing to provide DHE unless ffdhe groups were advertised has sense too

jvdsn commented:
@tomato42 I had to take a look at the code again, you're absolutely right. Sorry for the confusion; I updated my comment.

vdukhovni commented:
Since we'll still be supporting FFDHE in TLS 1.2 for the foreseeable future, I do think it makes sense to support the RFC7919 groups as a alternatives to the "auto" FFDHE groups (preëmpting them):

* When server's supported_groups includes a non-empty set of FFDHE groups matching the desired cipher security level.
* We're considering a DHE cipher, AND
    - The client's supported_groups includes no FFDHE code points (legacy client), OR
    - The client's supported_groups includes at least one server_supported FFDHE code point (RFC7919 client).

In terms of the text above, the RFC7919 groups would be chosen "prior" to "step 1", obviating the need to fall back to the IKE groups.

Note that in 3.5 the default group list includes only `ffdhe2048` and `ffdhe3072`, so these would be the only ones selected by default.

As far as requiring only matching groups, that's unlikely to be realistic in TLS 1.2, such configurations would be much too fragile for general deployment. It is not clear that a new feature to implement enforcement is justified.

Cc: @mattcaswell, @t8m

jvdsn commented:
> * The client's supported_groups includes no FFDHE code points (legacy client), OR

I'm trying to understand the use case for this. Right now, the code in this PR only uses shared groups advertised by client and supported by the server. If the server does not advertise any groups at all (i.e., does not include supported_groups), or only advertises EC curves, the code still falls back to "step 1" described above. That is, the MODP groups would be used. You're saying in that case you still want to use RFC 7919 groups?

Let's say there's a few possibilities, assuming the server has RFC 7919 groups configured:
a) Client does not advertise any supported_groups -> choose any RFC 7919 group appropriate for the SSL security level?
b) Client advertises supported_groups but only EC curves -> choose any RFC 7919 group appropriate for the SSL security level?
c) Client advertises supported_groups, including RFC 7919 groups, but there's no overlap (appropriate for the SSL security level) -> go to "step 1"?
d) Client advertises supported_groups, including RFC 7919 groups, and there's overlap (appropriate for the SSL security level) -> choose any RFC 7919 group appropriate for the SSL security level

Is that what you're looking for @vdukhovni?

vdukhovni commented:
> > * The client's supported_groups includes no FFDHE code points (legacy client), OR
> 
> I'm trying to understand the use case for this. Right now, the code in this PR only uses shared groups advertised by client and supported by the server. If the server does not advertise any groups at all (i.e., does not include supported_groups), or only advertises EC curves, the code still falls back to "step 1" described above. That is, the MODP groups would be used. You're saying in that case you still want to use RFC 7919 groups?
> 
> Let's say there's a few possibilities, assuming the server has RFC 7919 groups configured:
>
> 1. Client does not advertise any supported_groups -> choose any RFC 7919 group appropriate for the SSL security level?
> 2. Client advertises supported_groups but only EC curves -> choose any RFC 7919 group appropriate for the SSL security level?
> 3. Client advertises supported_groups, including RFC 7919 groups, but there's no overlap (appropriate for the SSL security level) -> go to "step 1"?
> 4. Client advertises supported_groups, including RFC 7919 groups, and there's overlap (appropriate for the SSL security level) -> choose any RFC 7919 group appropriate for the SSL security level
> 
> Is that what you're looking for @vdukhovni?

Almost, but I think that, in "3", if a client actually advertises specific RFC7919 groups, but there's no overlap, we should not choose a DHE cipher, because there's no supported combination. So if a client asks only for ffedhe4096 or higher, but we only support ffdhe2048 and ffdhe3072, we use ECDHE or fail.

This is my perpective, but this is a complex topic with various tradeoffs, so I'd like to hear from others. The basic idea is to go with RFC7919 as much as reasonably possible.

jvdsn commented:
> Almost, but I think that, in "3", if a client actually advertises specific RFC7919 groups, but there's no overlap, we should not choose a DHE cipher, because there's no supported combination. So if a client asks only for ffedhe4096 or higher, but we only support ffdhe2048 and ffdhe3072, we use ECDHE or fail.

That is what RFC 7919 prescribes, but it would not be backwards compatible I guess:

    If a compatible TLS server receives a Supported Groups extension from
    a client that includes any FFDHE group (i.e., any codepoint between
    256 and 511, inclusive, even if unknown to the server), and if none
    of the client-proposed FFDHE groups are known and acceptable to the
    server, then the server MUST NOT select an FFDHE cipher suite.

vdukhovni commented:
> > Almost, but I think that, in "3", if a client actually advertises specific RFC7919 groups, but there's no overlap, we should not choose a DHE cipher, because there's no supported combination. So if a client asks only for ffedhe4096 or higher, but we only support ffdhe2048 and ffdhe3072, we use ECDHE or fail.
> 
> That is what RFC 7919 prescribes, but it would not be backwards compatible I guess:
> 
> ```
> If a compatible TLS server receives a Supported Groups extension from
> a client that includes any FFDHE group (i.e., any codepoint between
> 256 and 511, inclusive, even if unknown to the server), and if none
> of the client-proposed FFDHE groups are known and acceptable to the
> server, then the server MUST NOT select an FFDHE cipher suite.
> ```

I'm inclined to follow that text, since legacy TLS 1.2 clients won't send such groups, and those do, are presumably ones that follow RFC7919, and have reasonable expectation of conformance.
Of course some might only expect the RFC7919 groups to have been used in TLS 1.3, with TLS 1.2 unaffected, so we might need to give that some though, but as a starting premise I think that the behaviour is reasonable.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 61 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 123 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 154 days ago

vdukhovni commented:
@jvdsn Are you still inclined to see this PR through to adoption? It seems from the discussion that there are some tweaks remaining to be done. Or did I miss the associated code updates?

vdukhovni commented:
> @jvdsn Are you still inclined to see this PR through to adoption? It seems from the discussion that there are some tweaks remaining to be done. Or did I miss the associated code updates?

Three weeks ago I asked whether this PR is still likely to move forward. Personally, I do think it makes sense to extend support for RFC7919 in TLS 1.2, so I'd like to see this done.  But if we're giving up, the PR can be closed...

jvdsn commented:
@vdukhovni I just force pushed to clean up the branch (since the logic is quite different now) and split it up in 4 commits that should be relatively readable individually. The diff in `ssl/statem/statem_srvr.c` looks a bit odd because the indentation changed.

I also updated the PR description to explain what's going on now. I'll still have to fix the code style checks.

vdukhovni commented:
> Humour me a bit because I'm slow and dumb. Am I reading you that we would be compliant with section 4 of RFC7919 in this case?

I wouldn't say *slow*, but perhaps not quite as familiar with this dark corner of the protocol space, ...
 
> We are free to try whatever we like if the client does not send supported groups... ok.

Yes, an I'm advocating use of the more "modern" RFC7919 groups in preference to the dated MODP groups.

> Presumably, If the client offers RFC7919 FFDHE groups, we consider them, in whatever preference order, as per normal, if we both recognize them and find them acceptable. In this case these groups are participating in the overlap with the server. also fine.

Yes.

> When the client sends RFC7919 groups, as well as non FFDHE groups, and we overlap with the client only on non FFDHE groups _as long as we are only then choosing from the overlapping groups_ I think it's fine, we will only choose the non-FFDHE overlapping match, either in client or server preference order.

And the overlapping groups are then (TLS 1.2, so PQC out of scope) exclusively ECDHE groups and we'd need to not choose a `kDHE` cipher.

> What's not clear is when you say there is "neither client nor server support any RFC7919 groups" If I the server know I support RFC7919 groups (presumbaly I do), and the client sends me an RFC7919 group I find unacceptable, and two other things I don't match, what happens?

If the server's supported groups configuration  includes a non-empty set of RFC7919 groups, then, per the RFC, we assume these are the only ones it is willing to use, and then if a client sends a non-empty, non-overlapping subset of RFC7818 groups, again we avoid kDHE, because there's no overlap.

> Am I free to pretend this the same case as if the client sent no list and try it anyway?

I don't think so.  If the client wants `ffdhe8192` or bust, and the server is only willing to burn CPU on `ffdhe2048` or `ffdhe3072` then FFDHE is not used.

> RFC7919 says I MUST not. How are we determining "client supports it?"?
> 1) looking at the values in the supported groups extension to see if they sent one works,

Check for a non-empty list of client-supported 7919 groups. If this is the case, then these are the only ones acceptable to the client, otherwise, the server's choice is not restricted.

I hope that's somewhat more clear?

jvdsn commented:
@vdukhovni making that change causes more test failures (presumably those tests rely on the extension _not_ being sent for TLS 1.2), so I'd rather not do it in this PR. I may open a separate PR to enable RFC 7919 in TLS 1.2 on the client side.

jvdsn commented:
Code style issue is due to the indents that were already incorrect prior to this PR. Not sure what the preferred way is to resolve it.

vdukhovni commented on ssl/ssl_lib.c:
```diff
@@ -4607,7 +4607,8 @@ void ssl_set_masks(SSL_CONNECTION *s)
               || c->dh_tmp_cb != NULL
-              || c->dh_tmp_auto);
+              || c->dh_tmp_auto
+              || tls1_shared_group(s, -2, 1, 0) != 0);
```
The magic `-2` needs some explanation, or at least a clear macro name to signal what it means.
Even the `1` and `0` could be macros `USE_FFDHE` and `NO_EC` or better names, or perhaps:
`1 /*ffdhe*/, 0 /*ec*/`

vdukhovni commented on ssl/statem/statem_srvr.c:
```diff
@@ -2577,49 +2577,63 @@ CON_FUNC_RETURN tls_construct_server_key_exchange(SSL_CONNECTION *s,
-            pkdh = ssl_get_auto_dh(s);
-            if (pkdh == NULL) {
+        /* Get NID of appropriate shared FFDHE group */
+        group_id = tls1_shared_group(s, -2, 1, 0);
```
Again that magic `-2`.

vdukhovni commented on ssl/ssl_local.h:
```diff
@@ -2850,7 +2850,8 @@ __owur int tls1_get0_implemented_groups(int min_proto_version,
                                         STACK_OF(OPENSSL_CSTRING) *out);
-__owur uint16_t tls1_shared_group(SSL_CONNECTION *s, int nmatch);
+__owur uint16_t tls1_shared_group(SSL_CONNECTION *s, int nmatch, int ffdhe,
+                                  int ecdhe);
```
I rather think that since `ecdhe` here is anything other than `ffdhe`, a single non-boolean argument would be better, with three values:

* NON_FFDHE_GROUPS = 0
* FFDHE_GROUPS = 1
* ALL_GROUPS = 2

Or equivalent.

vdukhovni commented on test/sslapitest.c:
```diff
@@ -11209,10 +11223,229 @@ static int test_dh_auto(int idx)
+    case 0:
+        clientgroup = "ffdhe2048";
+        servergroup = "ffdhe3072";
+        ciphersuite = "DHE-RSA-AES128-SHA:AES128-SHA";
```
For TLS 1.2 with RSA key exchange, should we perhaps specify `AES128-SHA256` instead?

vdukhovni commented on test/sslapitest.c:
```diff
@@ -11209,10 +11223,229 @@ static int test_dh_auto(int idx)
+        goto end;
+    }
+
+    if (TEST_int_eq(TLS1_CK_DHE_RSA_WITH_AES_128_SHA,
```
In which case, `TLS1_CK_DHE_RSA_WITH_AES_128_SHA256 ` here?

vdukhovni reviewed (commented):
LGTM, many thanks.  But before I mark the PR approved, I'd like to take it for a spin in my Postfix server first, so I'll be back before long with the final verdict, or less likely some final nits.

jvdsn commented:
@vdukhovni do you have any update here?

vdukhovni commented:
> The updated PR seems to pass every test I throw at it, so I'm good with it.

Thanks for checking. Now we just need some independent reviewers.

vdukhovni commented:
Rebased to absorb updates to `CHANGES.md` and removal of SSLv3.

jvdsn commented on test/sslapitest.c:
```diff
@@ -11326,7 +11326,9 @@ static int test_dh_auto(int idx)
-        || !TEST_true(SSL_set_cipher_list(clientssl, ciphersuite)))
+        || !TEST_true(SSL_set_cipher_list(clientssl, ciphersuite))
+        /* This is required so the server does not use RFC7919 groups */
+        || !TEST_true(SSL_set1_groups_list(clientssl, "xorgroup")))
```
Otherwise tests fail in no-ec builds.

mattcaswell commented on ssl/statem/extensions_clnt.c:
```diff
@@ -136,43 +137,121 @@ EXT_RETURN tls_construct_ctos_srp(SSL_CONNECTION *s, WPACKET *pkt,
+            return 1;
+    }
     return 0;
 }
```
This comment is still outstanding

vdukhovni commented on ssl/statem/extensions_clnt.c:
```diff
@@ -167,9 +167,12 @@ static ossl_inline int is_ffdhe_group(const TLS_GROUP_INFO *ginfo)
-static int negotiate_ecdhe(SSL_CONNECTION *s, int ptformats, int min_version,
-    int max_version)
+static int negotiate_dhe(SSL_CONNECTION *s, int ptformats, int ecdhe, int ffdhe,
+    int min_version, int max_version)
```
There are only three valid combinations of `ptformats`, `ecdhe`, and `ffdhe` on entry to this function:

1. `ffdhe = 1`, `ecdhe = ptformats = 0`: an FFDHE check.
2. `ecdhe = 1`, `ffdhe = ptformats = 0`: an ECDHE check
3. `ptformats = 1`, `ecdhe = ffdhe = 0`: a point formats check (perhaps this could instead have `ecdhe = 1`, but that's an unimportant detail)

Whereas the three separate boolean parameters make for 8 combinations, which would lead to bugs if misused.
Therefore, these need to become just one `enum` variable with three possible states:

```c
typedef enum {
    ffdhe_check,
    ecdhe_check,
    ptfmt_check
} dhe_check_t;
```
You should instead pass a value for that, and use a switch or if statements as appropriate.

This of course assumes that Matt still prefers the unified form.  I was the one who chose to separate these to make each branch simpler, so I'll not weigh in on which is better, either will do, my preference was clear, but not emphatic.

jvdsn replied:
@vdukhovni I just pushed an updated commit. Essentially, I took the previous logic from the cases in your commit (https://github.com/openssl/openssl/pull/24551/files/3b4339f4808cd52a7601f2cd0bcd2b9ab32be814#diff-f0b1e90a60a50cb755b7b0dad66932d75595f9a280cd6a6f209f59a344a9d126) and explicitly added it based on the enum declaration. Below is the logic I used:

(negotiate_ecdhe) `ptformats = 1` aka `ptfmt_check` ->
```c
(alg_k & (SSL_kECDHE | SSL_kECDHEPSK))
            || (alg_a & SSL_aECDSA)
```
```c
tls_valid_group(s, ctmp, min_version, max_version, NULL, &ginfo)
            && is_ecdhe_group(ginfo)
            && tls_group_allowed(s, ctmp, SSL_SECOP_CURVE_SUPPORTED)
```

(negotiate_ecdhe) `ptformats = 0` aka `ecdhe_check` ->
```c
(alg_k & (SSL_kECDHE | SSL_kECDHEPSK))
            || (alg_a & SSL_aECDSA)
            || (1
                && (dtls ? DTLS_VERSION_GT(c->min_dtls, DTLS1_2_VERSION)
                         : (c->min_tls > TLS1_2_VERSION)))
```
```c
tls_valid_group(s, ctmp, min_version, max_version, NULL, &ginfo)
            && is_ecdhe_group(ginfo)
            && tls_group_allowed(s, ctmp, SSL_SECOP_CURVE_SUPPORTED)
```

(negotiate_ffdhe) aka `ffdhe_check` ->
```c
alg_k & (SSL_kDHE | SSL_kDHEPSK))
            || (dtls ? DTLS_VERSION_GT(c->min_dtls, DTLS1_2_VERSION)
                     : (c->min_tls > TLS1_2_VERSION))
```
```c
tls_valid_group(s, ctmp, min_version, max_version, NULL, &ginfo)
            && is_ffdhe_group(ginfo)
            && tls_group_allowed(s, ctmp, SSL_SECOP_TMP_DH))
```

This is then simplified a bit (in particular the `tls_valid_group` call is separated out).

mattcaswell commented:
Ping @vdukhovni - please take another look.

Also looks like there is a conflict int the CHANGES.md file so a rebase is needed.

vdukhovni commented on ssl/statem/extensions_clnt.c:
```diff
@@ -153,30 +186,78 @@ static int use_ecc(SSL_CONNECTION *s, int min_version, int max_version)
 
+        if (tls_valid_group(s, ctmp, min_version, max_version, NULL, &ginfo)
+            && is_ffdhe_group(ginfo)
+            && tls_group_allowed(s, ctmp, SSL_SECOP_CURVE_SUPPORTED))
```
This is now inconsistent with the check in the loop that constructs the extension, in which `SSL_SECOP_CURVE_SUPPORTED` is passed to `tls_group_allowed()` for both EC and FFDHE groups.
The built default security callback does not care, and checks the `bits` value. But we should perhaps either do the same there, or use `SSL_SECOP_CURVE_SUPPORTED` for both types of groups where *curve* is just understood to mean any kex group (as is I guess the case now also with ML-KEM, ...).

So I'm somewhat leaning towards `SSL_SECOP_CURVE_SUPPORTED` (and perhaps a doc update to make this clear).  That third argument is basically the always one of:
```
SSL_SECOP_CURVE_SUPPORTED
SSL_SECOP_CURVE_SHARED
SSL_SECOP_CURVE_CHECK
```
where by *curve* we now mean **group**.

vdukhovni commented on ssl/statem/extensions_clnt.c:
```diff
@@ -244,19 +305,24 @@ EXT_RETURN tls_construct_ctos_supported_groups(SSL_CONNECTION *s, WPACKET *pkt,
+                &ginfo)
+            || (!use_ecdhe && is_ecdhe_group(ginfo))
+            || (!use_ffdhe && is_ffdhe_group(ginfo))
+            || !tls_group_allowed(s, ctmp, SSL_SECOP_CURVE_SUPPORTED))
```
A comment may be approbate here to note that `SSL_SECOP_CURVE_SUPPORTED` covers all key exchange *groups* (EC, FFDHE and KEMs).
See: https://github.com/openssl/openssl/pull/24551/changes#r2710810904

vdukhovni commented on ssl/statem/extensions_clnt.c:
```diff
@@ -244,19 +305,24 @@ EXT_RETURN tls_construct_ctos_supported_groups(SSL_CONNECTION *s, WPACKET *pkt,
+            SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+            return EXT_RETURN_FAIL;
         }
+        if (okfortls13 && max_version == TLS1_3_VERSION)
```
Food for thought: Should that be `>=` (though who knows whether TLS 1.4 will happen, and what it might be like).
And what about DTLS here?

mattcaswell replied:
There are so many places where we would have to consider support for TLSv1.4 if such a thing ever happened that I don't think we should worry too much about it here.

As for DTLS, this is only relevant for DTLSv1.3 which we don't support in master yet. The feature/dtls-1.3 code looks a little different and already handles this case:

https://github.com/openssl/openssl/blob/1f0c23c4c8b47a46f18ddebd5c3ac698b5316e8f/ssl/statem/extensions_clnt.c#L263-L264

There will be merge conflicts as a result of this change when the DTLSv1.3 branch gets merged to master. But that is not a problem for this PR.

vdukhovni commented on ssl/t1_lib.c:
```diff
@@ -1050,9 +1055,28 @@ uint16_t tls1_shared_group(SSL_CONNECTION *s, int nmatch)
+             * If the caller is only interested in FFDHE groups or the group is
+             * not allowed, ignore it
+             */
+            if (groups == TLS1_GROUPS_FFDHE_GROUPS
```
This seems wrong, we should still be checking "group allowed" for FFDHE groups even if the request is for `all groups`. Another aspect of the same question.

vdukhovni commented on ssl/t1_lib.c:
```diff
@@ -1745,8 +1769,14 @@ int tls1_check_group_id(SSL_CONNECTION *s, uint16_t group_id,
+    } else {
+        if (!tls_group_allowed(s, group_id, SSL_SECOP_CURVE_CHECK))
+            return 0;
+    }
```
If not unifying the two cases (preferred if @mattcaswell agrees), then another opportunity to use the predicate.

vdukhovni commented on test/sslapitest.c:
```diff
@@ -11349,6 +11350,228 @@ static int test_dh_auto(int idx)
+
+    if (!TEST_true(SSL_set_dh_auto(serverssl, 1))
+        || !TEST_true(SSL_set_min_proto_version(serverssl, TLS1_2_VERSION))
+        || !TEST_true(SSL_set_max_proto_version(serverssl, TLS1_2_VERSION))
```
As expected, the tests are for TLS 1.2.

vdukhovni commented:
@mattcaswell Please take a look at my (almost) final review comments, I'd like to "hear" your thoughts.

vdukhovni reviewed (approved):
LGTM.
